Periodic boundary conditions for Linear and Constant

jaemolihm opened the pull request:
As discussed in #326, this PR enables periodic linear and constant interpolation without adding additional points.

The major change is that `Linear` and `Constant` are now subtypes of DegreeBC. The boundary conditions can only be `Throw(OnGrid())` or `Periodic(OnCell())` because others does not make sense or are redundant.

Existing constructors keeps working. Periodic constant and linear interpolations can be constructed similarly to cubic ones, e.g. `Linear(Periodic(OnCell()))`.

# Examples

### Example 1 (taken from https://github.com/JuliaMath/Interpolations.jl/issues/326#issuecomment-505220728)
```Julia
using Interpolations
using PyPlot
fig, axes = subplots(1, 2, figsize=(8, 4))
f(x) = sin(x)
xs = range(0, stop=2pi, length=10)[1:end-1] # note that we don't want the overlapping point at 2pi
A = f.(xs);
for i in 1:2
    sca(axes[i])
    if i == 1
        title("old")
        itp = interpolate(A, BSpline(Linear()), OnGrid()) # Old
    else
        title("new")
        itp = interpolate(A, BSpline(Linear(Periodic(OnCell())))) # New
    end
    stp = scale(itp, xs) # re-scale to the actual domain
    etp = extrapolate(stp, Periodic()) # extrapolate periodically

    xs_long = range(-2pi, stop=4pi, length=100)
    A_long = [etp[x] for x in xs_long]
    plot(xs, A, "o", label="Data")
    plot(xs_long, A_long, label="Extrapolated")
    plot(xs_long, sin.(xs_long), label="sin(x)")
    legend(loc="lower left")
end
display(gcf()); clf()
close(fig)
```

![image](https://user-images.githubusercontent.com/32537613/119788410-156ee000-bf0d-11eb-8445-08ec8ce874f3.png)

### Example 2

```Julia
using Interpolations
using PyPlot
f(x) = sin(2π * x + 0.5)
df(x) = 2π * cos(2π * x + 0.5)
N = 5
x = range(0, 1, length=N+1)[1:end-1]
y = f.(x)
degs = reshape([
    Constant(),
    Constant{Previous}(),
    Constant{Next}(),
    Linear(),
    Constant(Periodic(OnCell())),
    Constant{Previous}(Periodic(OnCell())),
    Constant{Next}(Periodic(OnCell())),
    Linear(Periodic(OnCell()))
    ], 4, 2)
fig, axes_ = subplots(size(degs)..., figsize=(12, 8), sharex=true, sharey=true)
for (i, deg) in enumerate(degs)
    sca(axes_[i])
    itp = interpolate(y, BSpline(deg));
    zero_to_one = range(0, 1, length=N+1)[1:end-1]
    stp = scale(itp, zero_to_one);
    etp = extrapolate(stp, Periodic());
    xx = range(-0.5, 2.5, step=0.001);
    plot(xx, etp.(xx), "C0-", label="Interpolated")
    plot(x, y, "ko", label="Data")
    plot(xx, f.(xx), "k--", label="Exact")
    # Plot derivatives
    if i in [4, 8]
        plot(xx, Interpolations.gradient.(Ref(etp), xx) ./ 5, "C1-")
        plot(xx, df.(xx) ./ 5, "y--")
    end
    title("$deg")
    if i == 1
        legend(loc="lower right")
    end
end
display(gcf()); clf()
close(fig)
```

Plots on the right side are the new features. The orange (yellow) lines are the interpolated (exact) derivatives.

![image](https://user-images.githubusercontent.com/32537613/119795412-67b2ff80-bf13-11eb-8a2e-2069a8fe0a48.png)

# TODO

- [x] Documentation
- [x] Add tests
- [x] Add convenience constructor `Linear(Periodic()) = Linear(Periodic(OnCell()))`?

closes #326

mkitti commented:
@jaemolihm Sorry I'm a little confused. Is this ready for review or work-in-progress?

jaemolihm commented:
Sorry for the confusion. The code is ready for review. The WIP is for the documentation and tests about the new feature. Also sorry for the delay on adding the docs and tests. I'll try to do it soon.

mkitti commented on src/b-splines/constant.jl:
```diff
@@ -5,8 +5,24 @@ struct Nearest <: ConstantInterpType end
 struct Next <: ConstantInterpType end
 
-struct Constant{T<:ConstantInterpType} <: Degree{0} end
+# struct Constant{T<:ConstantInterpType} <: Degree{0} end
+# Constant() = Constant{Nearest}()
```
You can remove this comment.

jaemolihm replied:
Done

mkitti reviewed (changes requested):
Other than cleaning up some comments, the changes look fine to me. I will see if I can help you out with some tests and documentations.

jaemolihm commented:
I added tests on periodic interpolation (the `lbounds` and `ubounds` are extended by 0.5) as well as periodic extrapolation. I also added convenience constructors, e.g. `Linear(Periodic()) = Linear(Periodic(OnCell()))`

jaemolihm commented:
I added a subsection explaining periodic extrapolation in the docs.

This PR is no more WIP and is fully ready for review. Sorry for the long delay.

jaemolihm commented:
Thanks or the comment. I changed the printing of `Linear()` and `Constant()` as you suggested.
``` julia
julia> Linear()
Linear()
julia> Linear(Throw(OnGrid()))
Linear()
julia> Linear(Periodic(OnCell()))
Linear(Periodic(OnCell()))
```

mkitti commented:
Could we add a test for `(deg::Linear)(gt::GridType)`?

mkitti commented:
We also need a test for `Base.show(io::IO, deg::Constant)`

mkitti commented on test/b-splines/constant.jl:
```diff
@@ -116,5 +116,53 @@
+        @test Constant() isa Constant{Nearest,Throw{OnGrid}}
+        @test Constant(Periodic()) === Constant(Periodic(OnCell()))
+        @test Constant(Periodic()) isa Constant{Nearest,Periodic{OnCell}}
+        for T in (Nearest, Previous, Next)
+            it = Constant{T}()
+            @test it isa Constant{T, Throw{OnGrid}}
+            it = Constant{T}(Periodic())
+            @test it isa Constant{T, Periodic{OnCell}}
+        end
```
```suggestion
        for T in (Nearest, Previous, Next)
            it = Constant{T}()
            @test it isa Constant{T, Throw{OnGrid}}
            @test "$it" == "Constant{$T}()"
            it = Constant{T}(Periodic())
            @test it isa Constant{T, Periodic{OnCell}}
            @test "$it" == "Constant{$T}(Periodic(OnCell()))"
        end
```

jaemolihm replied:
Done, thanks.

mkitti commented on src/b-splines/linear.jl:
```diff
@@ -1,4 +1,14 @@
+    bc::BC
+end
+
+(deg::Linear)(gt::GridType) = Linear(deg.bc(gt))
```
Do we actually need this method?
```julia
julia> L = Linear()
Linear()

julia> L(OnGrid())
Linear()

julia> L(OnCell())
ERROR: MethodError: no method matching Linear(::Throw{OnCell})
...
```

jaemolihm replied:
I agree, we do not need this method. (It was a leftover from copying from cubic.)

mkitti commented on test/b-splines/linear.jl:
```diff
@@ -61,4 +61,39 @@
+        @test Linear() === Linear(Throw(OnGrid()))
+        @test Linear() isa Linear{Throw{OnGrid}}
+        @test Linear(Periodic()) === Linear(Periodic(OnCell()))
+        @test Linear(Periodic()) isa Linear{Periodic{OnCell}}
+
```
If we need `(deg::Linear)(gt::GridType)`, test it here.

```suggestion
        @test Linear(Periodic()) isa Linear{Periodic{OnCell}}
        @test Linear() === Linear()(OnGrid())

```

jaemolihm replied:
I removed `(deg::Linear)(gt::GridType)`.

mkitti reviewed (changes requested):
Here's a few suggestions for the tests. I'm not sure if we need `(deg::Linear)(gt::GridType)`.

mkitti commented:
What would you think of syntax that works as follows?

```julia
julia> import Interpolations: Periodic, GridType

julia> Periodic(::Type{GTT}) where GTT <: Union{GridType,Nothing} = Periodic(GTT())
Periodic

julia> itp = interpolate([1., 2., 3., 2.], OnCell |> Periodic |> Linear |> BSpline)
4-element interpolate(::Vector{Float64}, BSpline(Linear(Periodic(OnCell())))) with element type Float64:
 1.0
 2.0
 3.0
 2.0

```

jaemolihm commented:
> ```julia
> Periodic(::Type{GTT}) where GTT <: Union{GridType,Nothing} = Periodic(GTT())
> ```

I guess this idea is from #401. I agree that piping constructor is more concise than parenthetical ones. Also, this issue is not limited to `Constant` or `Linear`. It applies to all `BSpline`.

As I understand, this new constructor is needed to make the first argument `OnGrid`, not `OnGrid()`. But I'm not sure that this is really desirable. All other constructors (like `Linear` or `BSpline`) accepts instances, not types, as arguments. So, I think `OnCell() |> Periodic |> Linear |> BSpline` is good enough.

For example, such a problem also occur when the starting point is `Periodic`, not `OnCell`. Then, one would also need to define `Cubic(::Type{BC}) where BC <: BoundaryCondition`, and also same thing for `Constant`, `Linear`, and `Quadratic`, which I think makes the code more complicated.
```julia
julia> Periodic() |> Cubic |> BSpline
BSpline(Cubic(Periodic()))

julia> Periodic |> Cubic |> BSpline
ERROR: MethodError: no method matching Cubic(::Type{Periodic})
Closest candidates are:
  Cubic(::BC) where BC<:Interpolations.BoundaryCondition at /home/jmlim/.julia/dev/Interpolations/src/b-splines/cubic.jl:2
Stacktrace:
 [1] |>(x::Type, f::Type{Cubic})
   @ Base ./operators.jl:858
 [2] top-level scope
   @ REPL[39]:1
```

(I do not know the Julia language design very well, so I may be wrong.)

johnnychen94 commented on src/b-splines/constant.jl:
```diff
@@ -5,8 +5,25 @@ struct Nearest <: ConstantInterpType end
-struct Constant{T<:ConstantInterpType} <: Degree{0} end
-Constant() = Constant{Nearest}()
+struct Constant{T<:ConstantInterpType,BC<:Union{Throw{OnGrid},Periodic{OnCell}}} <: DegreeBC{0}
+    bc::BC
```
Is `bc` stateful here? Or could we always use `BC()` to construct it at runtime?

jaemolihm replied:
https://github.com/JuliaMath/Interpolations.jl/blob/a8370bd2b4ef77cfe689c55c3ed217ab528f69c1/src/b-splines/b-splines.jl#L142
I am not sure I understand the question, but I think since the above line uses `deg.bc`, so `bc` should be a field.
